### ISSUE TITLE
[Proposal] Support both empty array and field-omitted for fields like remote_addrs

### ISSUE THREAD

**Issue by juzhiyuan** [apache/apisix#2924]
## Background

When creates a route in APISIX, the optional fields are either passed or not passed, and for array-type parameters, passing an empty array or not passing the field should serve the same purpose.

APISIX currently intercepts empty array data, which results in the Dashboard requiring additional processing of certain fields in the data before sending requests, such as `remote_addrs` field.

```ts
// Omit this field if matched this rule
data.remote_addrs?.filter(Boolean).length === 0 ? 'remote_addrs' : '',
```

## Solution

Omit that field and use an empty array are quiet 2 things indeed, but the frontend also shouldn't care about which array-like fields should be omitted either, so I would suggest supporting an empty array in APISIX (Data Plane) to prevent those hack codes in the dashboard.

## Reference

- https://github.com/apache/apisix-dashboard/pull/948#pullrequestreview-542517940

cc @membphis @nic-chen  @spacewander @liuxiran @tokers @LiteSun

**Comment by spacewander** [comment]
Put my disagreement here:
https://github.com/apache/apisix-dashboard/pull/948#discussion_r533932864

Yes, but the backends care. This change just shift the duty to the backend.
There are three problems:
1. We have two backends. It is hard to sync the validation. In the future, we may have other backends like envoy, should we do the same again?
2. jsonschema treat `empty array` and `omit array field` different. We have to write code manually twice.
3. What if one day we want to distinguish `[]` from `nil`?

In short, type matters. If `[]` and `nil` are two different types, we should distinguish them from the source.

**Comment by juzhiyuan** [comment]
After some thought, I agree with your opinion on the data plane, maybe as the bridge between frontend and data source, the manager-api would be more reasonable to take this responsibility IMO.

**Comment by juzhiyuan** [comment]
A more detailed documentation for manager-api is needed emm

**Comment by tokers** [comment]
Agree with @spacewander 's point, we should differentiate empty array and nil, we may need to apply some default manner when a field is empty, treating it as same as empty array is just one situtation.

**Comment by liuxiran** [comment]
May I ask  in what scenario we're going to distinguish between [] and nil?  @spacewander 

From the FE point of view, which is closer to the meaning of data expression than to semantics at the code level,  both the field empty and nil all indicate that the field is not configured, so we would suggest that treate the two types of data equally.

Data transformation between user data and underlying data is inevitable, It's best that these two data are the same, if there is a disagreement, we may need an agreement to deal with it.

**Comment by spacewander** [comment]
@liuxiran 
`remote_addrs: []` and `remote_addrs: nil` are two different things in the backend code. The former one means: I want to match the remote_addrs, but the given addresses are empty. The latter one means: I don't want to match remote_addrs.

Even there is not problem today, we will be  bitten in the future if we don't take care about the type (Like the integer id 1 VS string id "1", and many other cases happened in my career).

**Comment by juzhiyuan** [comment]
> @liuxiran 
> `remote_addrs: []` and `remote_addrs: nil` are two different things in the backend code. The former one means: I want to match the remote_addrs, but the given addresses are empty. The latter one means: I don't want to match remote_addrs.
> 
> Even there is not problem today, we will be  bitten in the future if we don't take care about the type (Like the integer id 1 VS string id "1", and many other cases happened in my career).

This description makes sense to me, then what's your opinion? @membphis

**Comment by membphis** [comment]
`remote_addrs: []`

For users, why is an empty array set here? What is his/her expectation?

There are two result for `remote_addrs: []`:

1.  means never match
2. means always match

I vote to `2`. What is your choice?

**Comment by spacewander** [comment]
@membphis 
Don't make thing more complicate here. Why does `always match` rule make sense? We can reject empty array `remote_addrs` if needed.

**Comment by juzhiyuan** [comment]
The empty array is not used to `always match` when I first see it unless I noticed some tips.

**Comment by liuxiran** [comment]
>remote_addrs: [] and remote_addrs: nil are two different things in the backend code. The former one means: I want to match the remote_addrs, but the given addresses are empty. The latter one means: I don't want to match remote_addrs.
Even there is not problem today, we will be bitten in the future if we don't take care about the type (Like the integer id 1 VS string id "1", and many other cases happened in my career).

Thanks for @spacewander's instructions, it gives me more insight into the function.
> 1. means never match
> 2. means always match

Without any explanations I will vote to 1,  never match.

BTW, Let me explain why there is an empty array here.

- to configure data, FE usually choose **form** as a way to interact with users. 

- we will all the configuration items so that user can config all items they want.

- for the **optional items**, use can not fill any thing, that means user do not customize this item,  for these unconfigured options,  it **will be an empty Array, or empty Object**.

Therefore, the data on the FE is loyal to the interaction needs by default, while the underlying data(APISIX) is loyal to the technology itself. It's normal to have differences.

> maybe as the bridge between frontend and data source, the manager-api would be more reasonable to take this responsibility IMO

As @juzhiyuan said, when we need to convert data, manager-api would be more reasonable to take this responsibility. This allows the FE to focus on UE and APISIX to focus on function.

Premisely, we have to reach an agreement on how to guide the user to operate.

**Comment by moonming** [comment]
I agreed with @spacewander and @liuxiran 
manager API should do the job between UE and APISIX.
We can not just do all things in DP.
Yes, APISIX should deny empty array as input.

**Comment by juzhiyuan** [comment]
ok, then if there are no other opinions in 24 hours, I will transfer this issue to dashboard.

**Comment by membphis** [comment]
pls confirm:

If no "remote address" is filled, what data will be sent from FE.

1. no field "remote address"
2. empty array, and manager api will delete field "remote address"

**Comment by liuxiran** [comment]
> pls confirm:
> 
> If no "remote address" is filled, what data will be sent from FE.
> 
> Two cases:
> 
> 1. no field "remote address"
> 2. empty array, and manager api will delete field "remote address"

2. empty array

**Comment by juzhiyuan** [comment]
And manager api will not omit that field as far as i know.

**Comment by membphis** [comment]
> pls confirm:
> 
> If no "remote address" is filled, what data will be sent from FE.
> 
> Two cases:
> 
> 1. no field "remote address"
> 2. empty array, and manager api will delete field "remote address"

I want to confirm what is the final way after we fixed this issue. @juzhiyuan @liuxiran

**Comment by juzhiyuan** [comment]
I would prefer hacking this issue on the manager api side, for it's the bridge between fe and data store.

**Comment by liuxiran** [comment]
> I would prefer hacking this issue on the manager api side, for it's the bridge between fe and data store.

the same with @juzhiyuan

**Comment by juzhiyuan** [comment]
Then what's your opinion? @nic-chen @gxthrj any suggestion here?

**Comment by membphis** [comment]
> I would prefer hacking this issue on the manager api side, for it's the bridge between fe and data store.

I don't like this method, which means that the data saved in etcd is not the same as the data requested by the Admin user.

Here is an example:

```
# request with `remote_addrs`
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/hello",
    "remote_addrs": []
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:8080": 1
        }
    }
}'

# when we fetch route, there is no `remote_addrs`
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1'
{
    "uri": "/hello",
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:8080": 1
        }
    }
}'
```

And if we change to this way, the Admin API needs to process it in the same way too. We need to keep this hack way both in `Admin API` and `Manager API`.

As we talked about before, if Admin API rejects empty array `"remote_addrs": []`, then the Manager API should reject empty array too. Keep the same is important.

**Comment by membphis** [comment]
This is dangerous if the input and output of the API are not the same.

If there is logic to modify user input data, it will increase the difficulty of troubleshooting.

**Comment by nic-chen** [comment]
> Then what's your opinion? @nic-chen @gxthrj any suggestion here?

In the past, in order to improve the user experience, the logic of UE did not completely match APISIX itself, so the Manager API stores the data submitted by the user in mysql, and the data submitted to APISIX is formatted, which is reasonable. 

But now Manager API operates ETCD directly, and there is no intermediate storage layer, so it is not suitable for such conversion.

As @membphis  said, input and output should be consistent.

Such hackings may be known to everyone involved in the discussion, but others and newcomers will feel confused.

**Comment by juzhiyuan** [comment]
Agree with @spacewander that APISIX, as a direct operator of etcd, should not do ambiguous things. Nil is different from empty arrays.

Although Manager API also operates on APISIX, it is not APISIX, it's a bridge for frontend.

Take remote_addrs as an example, just pass an empty array AND respond to an empty array is fine. The frontend wants to ensure that this field is ALWAYS an array no matter sending requests or getting response from API, not null or undefined, or the frontend has to do such hacks for every array like field.

**Comment by juzhiyuan** [comment]
Ok fine, let me explain what happened here.

The frontend now has to include some hacking codes for array like fields, this change lead to 2 issues:

1. For users, they may find the data has been modified before sending requests, why?  what happened? Though it's ok to use that, and no errors would happen, it's a little strange IMO.

2. For developers, we have to leave multiple comments to prevent from confusing for newcomers, or it's too tricky. 

We shouldn't modify apisix's logics according to comments above.

We don't want to modify manager api's logics to prevent from more confusing.

We shouldn't add more hacking codes to frontend, and don't have to know what's happened in backend. 

Then what should we do? 

Fallback to docs, is it clear enough? 🤔 

1. Keep adding hacking codes to frontend, but should support a more clear and semantic docs, it's for developers. The frontend developers only focus on docs, they could develop without having to ask issues for api support. 

2. Hacking in manager api, this will take differences from apisix indeed, but will users use apisix directly or manager api directly? The nil and empty array are really different for users? 🤔 I'm not sure, if yes, keep adding hacking codes to frontend is acceptable currently. 

As for docs, after deprecating admin api, then the docs would call Docs for manager-api, serving for both developers(call api directly) and frontend users.

**Comment by membphis** [comment]
@nic-chen Your reason is not good. We should discuss the API design first, it has nothing to do with what type of storage is used.

>  The nil and empty array are really different for users? 🤔 

Yes, they are different for users now.

**Here is the todo list(from the discussion)**

* `Admin API`: reject request data if it contains an empty array, eg: `"remote_addrs": []`. `Code + Doc`
* `Manager API`: same as `Admin API`.
* `Dashbaord`: keep adding hacking codes to the frontend. `Code`

The field `hosts`, `uris` and `methods`, they have the same issue, we should fix them in the same way too.

**Comment by gxthrj** [comment]
I think `manager-api` is no longer a bridge between APISIX and dashboard, it is a backend for dashboard, and data is saved in `etcd`. Although `etcd` data is only used for APISIX, it can also be used for other services .

If the data plane thinks that the empty array is inconsistent with nil, then we should truthfully reflect these two data structures to `etcd`.

I don't want the `manager-api` to do too much judgment logic to adapt . These adaptations are endless.

In addition, I hope to keep the  behavior of `manager-api` and `Admin API` consistent to avoid troubles when migrating `Admin API` in the future.

**Comment by liuxiran** [comment]
> We should discuss the API design first

I'd like to move on from this point.

Should API parameters be designed to be **highly consistent with code logic**, or to **maximize user convenience**?

I perfer the latter one, because people who use an API can ignore how it's implemented behind it. So when we encapsulate API, I think it is necessary for us to do some data conversion for the better use of the interface by users.

Just take `remote_addr` and `remote_addrs` as an example. When user what to config remote_addr

- config one remote_addr

- config multiple remote_addrs

- do not config remote_addr 

what we're doing now is  that we have provided three transmission parameters to the users

- when you want to config one remote_addr, you should config `remote_addr`

- when you want to config multiple remote_addrs, you should config `remote_addrs`

- when you do not want to config remote_addr (one or multiple), you should omit this config item.

would it be better that we only give one `remote_addrs` instead of all above? and that will turn to

- when you want to config one remote_addr, you can use `remote_addrs: [addr1]`

- when you want to config multiple remote_addrs, you can use `remote_addrs: [addr1, addr2, ....]`

- when you do not want to config remote_addr, you can just config `remote_addrs: []` or omit it I think they all ok.

**Leaving the simple to the user, we need to use code to translate user requirements into better application requirements**

**Comment by moonming** [comment]
> > We should discuss the API design first
> 
> 
> 
> I'd like to move on from this point.
> 
> 
> 
> Should API parameters be designed to be **highly consistent with code logic**, or to **maximize user convenience**?
> 
> 
> 
> I perfer the latter one, because people who use an API can ignore how it's implemented behind it. So when we encapsulate API, I think it is necessary for us to do some data conversion for the better use of the interface by users.
> 
> 
> 
> Just take `remote_addr` and `remote_addrs` as an example. When user what to config remote_addr
> 
> 
> 
> - config one remote_addr
> 
> 
> 
> - config multiple remote_addrs
> 
> 
> 
> - do not config remote_addr 
> 
> 
> 
> what we're doing now is  that we have provided three transmission parameters to the users
> 
> 
> 
> - when you want to config one remote_addr, you should config `remote_addr`
> 
> 
> 
> - when you want to config multiple remote_addrs, you should config `remote_addrs`
> 
> 
> 
> - when you do not want to config remote_addr (one or multiple), you should omit these config item.
> 
> 
> 
> would it be better that we only give one `remote_addrs` instead of all above? and that will turn to
> 
> 
> 
> - when you want to config one remote_addr, you can use `remote_addrs: [addr1]`
> 
> 
> 
> - when you want to config multiple remote_addrs, you can use `remote_addrs: [addr1, addr2, ....]`
> 
> 
> 
> - when you do not want to config remote_addr, you can just config `remote_addrs: []` or omit it I think they are all ok.
> 
> 
> 
> **Leaving the simple to the user, we need to use code to translate user requirements into better application requirements**
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

agree with you

**Comment by moonming** [comment]
> I think `manager-api` is no longer a bridge between APISIX and dashboard, it is a backend for dashboard, and data is saved in `etcd`. Although `etcd` data is only used for APISIX, it can also be used for other services .
> 
> 
> 
> If the data plane thinks that the empty array is inconsistent with nil, then we should truthfully reflect these two data structures to `etcd`.
> 
> 
> 
> I don't want the `manager-api` to do too much judgment logic to adapt . These adaptations are endless.
> 
> 
> 
> In addition, I hope to keep the  behavior of `manager-api` and `Admin API` consistent to avoid troubles when migrating `Admin API` in the future. 
> 
> 
> 
> 
> 
> 
> 
> 

so what is your proposal?

**Comment by gxthrj** [comment]
I have two suggestions: 
* Set `remote_addrs` not empty or omit it, because the meaning of empty array is inconsistent with nil.
* Keep the behavior of `manager-api` and `Admin API` consistent to avoid troubles when migrating Admin API in the future.

**Comment by membphis** [comment]
> would it be better that we only give one `remote_addrs` instead of all above? and that will turn to
> 
> * when you want to config one remote_addr, you can use `remote_addrs: [addr1]`
> * when you want to config multiple remote_addrs, you can use `remote_addrs: [addr1, addr2, ....]`

For dashboard: it already works like this now.

In the future, only `remote_addrs` will be needed, `remote_addr` will be removed.
We will delete `remote_addr` in one milestone(Maybe  `3.0`).
Similar issue, we'll remove the field `host`, `uri`.

We can create a new issue about remove field `remote_addr`.

**Comment by juzhiyuan** [comment]
````
when you do not want to config remote_addr, you can just config remote_addrs: [] or omit it I think they are all ok.
````

@liuxiran provides 3 cases, how about the last one? This is the main issue we meet currently.

**Comment by membphis** [comment]
@juzhiyuan @liuxiran @nic-chen @moonming 

I will close this issue and will create a new issue. I think we have finished the discussion, and here is the final result.

Please feel free to reopen this issue if you have any problems with this issue.

**Here is the todo list(from the discussion)**

* `Admin API`: reject request data if it contains an empty array, eg: `"remote_addrs": []`. `Code + Doc`
* `Manager API`: same as `Admin API`.
* `Dashbaord`: keep adding hacking codes to the frontend. `Code`

The field `hosts`, `uris` and `methods`, they have the same issue, we should fix them in the same way too.
